pkg/auth/providers/saml/saml_client.go

```diff
@@ -262,6 +295,47 @@ func (s *Provider) getSamlPrincipals(config *v32.SamlConfig, samlData map[string
 	return userPrincipal, groupPrincipals, nil
 }
 
+// FinalizeSamlLogout processes the logout obtained by the POST to /saml/slo from IdP
+// ATTENTION: Expects to be called after the logout response was validated.
+// See `ServeHTTP`, saml_handlers.go
+func (s *Provider) FinalizeSamlLogout(w http.ResponseWriter, r *http.Request) {
+
+	if relayState := r.Form.Get("RelayState"); relayState != "" {
+		// delete the cookie
+		s.clientState.DeleteState(w, r, relayState)
+	}
+
+	redirectURL := s.clientState.GetState(r, "Rancher_FinalRedirectURL")
+
+	s.clientState.DeleteState(w, r, "Rancher_UserID")
+	s.clientState.DeleteState(w, r, "Rancher_Action")
+	s.clientState.DeleteState(w, r, "Rancher_FinalRedirectURL")
+
+	r.ParseForm()
+	err := s.serviceProvider.ValidateLogoutResponseRequest(r)
+	if err != nil {
+		log.Debugf("SAML [FinalizeSamlLogout]: response validation failed: %v", err)
+		if parseErr, ok := err.(*saml.InvalidResponseError); ok {
+			log.Debugf("SAML RESPONSE: ===\n%s\n===\nSAML NOW: %s\nSAML ERROR: %s",
+				parseErr.Response, parseErr.Now, parseErr.PrivateErr)
+		}
+
+		// assumes a redirect url with a `?...` already in place.
+		redirectURL += "&errorCode=500&errorMsg=" + err.Error()
```

Comment:
It is probably a good idea to ensure the values in `errorMsg` are encoded correctly.

```suggestion
		rURL, err := url.Parse(redirectURL)
		...
		params := rURL.Query()
		params.Add("errorCode", "500")
		params.Add("errorMsg", err.Error())
		
		rURL.RawQuery = params.Encode()
```

then we would use `rURL.String()` instead of `redirectURL`.

Reply:
Looks sensible. Note that the `ServeHTTP` code path handling the ACS endpoint has similar code, should be changed analogously.